code/game/machinery/_machines_base/stock_parts/network_lock.dm

```diff
@@ -0,0 +1,145 @@
+
+/obj/item/stock_parts/network_lock/proc/get_default_access()
+	if(auto_deny_all)
+		return list("NO_PERMISSIONS_DENY_ALL")
+	return list()
+
+/obj/item/stock_parts/network_lock/examine(mob/user)
+	. = ..()
+	if(emagged && user.skill_check_multiple(list(SKILL_FORENSICS = SKILL_EXPERT, SKILL_COMPUTER = SKILL_EXPERT)))
+		to_chat(user, SPAN_WARNING("On close inspection, there is something odd about the interface. You suspect it may have been tampered with."))
+
+/obj/item/stock_parts/network_lock/attack_self(var/mob/user)
+	ui_interact(user)
+
+/obj/item/stock_parts/network_lock/ui_data(mob/user, ui_key)
+	var/list/data[0]
+	. = data
+
+	var/datum/extension/network_device/D = get_extension(src, /datum/extension/network_device)
+	if(!istype(D))
+		error = "HARDWARE FAILURE: NETWORK DEVICE NOT FOUND"
+		data["error"] = error
+		return
+	data["error"] = error
+	data += D.ui_data(user, ui_key)
+
+	var/datum/computer_network/network = D.get_network()
+	if(!network)
+		data["connected"] = FALSE
+		return
+	data["connected"] = TRUE
+	data["default_state"] = auto_deny_all
+	var/list/grants = list()
+	if(!network.access_controller)
+		return
+	for(var/datum/computer_file/data/grant_record/GR in network.access_controller.get_all_grants())
+		grants.Add(list(list(
+			"grant_name" = GR.stored_data,
+			"assigned" = (GR.stored_data in grants)
+		)))
+	data["grants"] = grants
+
+/obj/item/stock_parts/network_lock/OnTopic(mob/user, href_list, datum/topic_state/state)
+	. = ..()
+	if(.)
+		return
+	. = TOPIC_HANDLED
+
+	if(href_list["refresh"])
+		error = null
+		return TOPIC_REFRESH
+	var/datum/extension/network_device/D = get_extension(src, /datum/extension/network_device)
+	if(!D)
+		return
+
+	if(href_list["settings"])
+		D.ui_interact(user)
+		return TOPIC_REFRESH
+
+	if(href_list["allow_all"])
+		auto_deny_all = FALSE
+		return TOPIC_REFRESH
+
+	if(href_list["deny_all"])
+		auto_deny_all = TRUE
+		return TOPIC_REFRESH
+
+	if(href_list["remove_grant"])
+		grants -= href_list["remove_grant"]
+		return TOPIC_REFRESH
+
+	if(href_list["assign_grant"])
+		grants |= href_list["assign_grant"]
+		return TOPIC_REFRESH
+
+/obj/item/stock_parts/network_lock/ui_interact(mob/user, ui_key, datum/nanoui/ui, force_open, datum/nanoui/master_ui, datum/topic_state/state)
+	var/data = ui_data(user)
+	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if (!ui)
+		ui = new(user, src, ui_key, "network_lock.tmpl", capitalize(name), 380, 500)
+		ui.set_initial_data(data)
+		ui.open()
+		ui.set_auto_update(1)
+
+/obj/item/stock_parts/network_lock/CouldUseTopic(var/mob/user)
+	..()
+	if(LAZYLEN(interact_sounds) && CanPhysicallyInteract(user))
+		playsound(src, pick(interact_sounds), interact_sound_volume)
+
+/obj/item/stock_parts/network_lock/CanUseTopic()
+	return STATUS_INTERACTIVE
+
+/obj/item/stock_parts/network_lock/buildable
+	part_flags = PART_FLAG_HAND_REMOVE
+	matter = list(MAT_STEEL = 400, MAT_GLASS = 200)
```

Comment:
don't set matter values like that anymore
gotta be like
```
	material = MAT_STEEL
	matter = list(MAT_GLASS = MATTER_AMOUNT_REINFORCEMENT)
```
It'll generate automagically from that.